crates/composition/src/composition_ir.rs

```diff
@@ -0,0 +1,222 @@
+    pub(crate) input_objects: Vec<federated::InputObject>,
+
+    pub(crate) strings: StringsIr,
+    pub(crate) fields: Vec<FieldIr>,
+    pub(crate) union_members: BTreeSet<(subgraphs::StringId, subgraphs::StringId)>,
+}
+
+impl CompositionIr {
+    pub(crate) fn insert_enum(&mut self, enum_name: StringWalker<'_>) -> federated::EnumId {
+        let name = self.insert_string(enum_name);
+        push_definition_and_return_id(
+            &mut self.enums,
+            federated::Enum {
+                name,
+                values: Vec::new(),
+                composed_directives: Vec::new(),
+            },
+            federated::EnumId,
+            |id| {
+                self.field_type_kinds_map
+                    .insert(enum_name.id, federated::FieldTypeKind::Enum(id));
+            },
+        )
+    }
+
+    pub(crate) fn insert_interface(
+        &mut self,
+        iface_name: StringWalker<'_>,
+    ) -> federated::InterfaceId {
+        let name = self.insert_string(iface_name);
+        push_definition_and_return_id(
+            &mut self.interfaces,
+            federated::Interface {
+                name,
+                composed_directives: Vec::new(),
+            },
+            federated::InterfaceId,
+            |id| {
+                self.field_type_kinds_map
+                    .insert(iface_name.id, federated::FieldTypeKind::Interface(id));
+            },
+        )
+    }
+
+    pub(crate) fn insert_scalar(&mut self, scalar_name: StringWalker<'_>) {
+        let name = self.insert_string(scalar_name);
+        push_definition_and_return_id(
+            &mut self.scalars,
+            federated::Scalar {
+                name,
+                composed_directives: Vec::new(),
+            },
+            federated::ScalarId,
+            |id| {
+                self.field_type_kinds_map
+                    .insert(scalar_name.id, federated::FieldTypeKind::Scalar(id));
+            },
+        );
+    }
+
+    pub(crate) fn insert_input_object(
+        &mut self,
+        input_object_name: StringWalker<'_>,
+    ) -> federated::InputObjectId {
+        let name = self.insert_string(input_object_name);
+        push_definition_and_return_id(
+            &mut self.input_objects,
+            federated::InputObject {
+                name,
+                fields: Vec::new(),
+                composed_directives: Vec::new(),
+            },
+            federated::InputObjectId,
+            |id| {
+                self.field_type_kinds_map.insert(
+                    input_object_name.id,
+                    federated::FieldTypeKind::InputObject(id),
+                );
+            },
+        )
+    }
+
+    pub(crate) fn insert_object(&mut self, object_name: StringWalker<'_>) {
+        let name = self.insert_string(object_name);
+        let object = federated::Object {
+            name,
+            implements_interfaces: Vec::new(),
+            resolvable_keys: Vec::new(),
```

Comment:
As you can see there, we're not filling in some of the important data. I didn't want to do it in this PR to avoid making it huger.